libr/arch/p/arch_bf.c

```diff
@@ -0,0 +1,343 @@
+/* radare - LGPL - Copyright 2020 - pancake, nibble */
+
+#include <r_arch.h>
+#include <r_anal.h> // just to get the R_ANAL_OP_TYPE_xx
+
+// TODO: move into r_arch
+#define has(y) (opt & R_ARCH_OPTION_##y)
+
+#define BUFSIZE_INC 32
+
+static size_t countChar(const ut8 *buf, int len, const char ch) {
+	size_t i;
+	for (i = 0; i < len; i++) {
+		if (buf[i] != ch) {
+			break;
+		}
+	}
+	return i;
+}
+
+static int getid(const char ch) {
+	const char *keys = "[]<>+-,.";
+	const char *cidx = strchr (keys, ch);
+	return cidx? cidx - keys + 1: 0;
+}
+
+static bool decode(RArchSession *a, RArchInstruction *ins, RArchDecodeOptions opt) {
+	r_return_val_if_fail (a && ins, false);
+	int len;
+	const ut8 *buf = r_strbuf_getbin (&ins->data, &len);
+	ins->type = R_ANAL_OP_TYPE_NOP;
+	ins->size = 1;
+	ins->opid = getid (buf[0]);
+	const char ch = buf[0];
+
+	if (opt == R_ARCH_OPTION_SIZE) {
+		ins->size = countChar (buf, len, *buf);
+		return true;
+	}
+	const char *buf_asm = "nop"; // invalid instructions do nothing in bfvm
+	switch (ch) {
+	case '[':
+		ins->type = R_ANAL_OP_TYPE_CJMP;
+		buf_asm = "while [ptr]";
+		// TODO: split the esil from the anal logic. the esil string should find the [ by itself
+		if (true) { // has (ANAL) || has (ESIL)) {
+			ut8 mydata[512] = {0};
+			const char *data = (const char *)buf;
+			const char *p = data + 1;
+			int lev = 0, i = 1;
+			bool tried = false;
+			len--;
+			while (i < len && *p) {
+				if (*p == '[') {
+					lev++;
+				}
+				if (*p == ']') {
+					lev--;
+					if (lev == -1) {
+						ut64 dst = ins->addr + i + 1;
```

Comment:
Maybe extract these processing in a separate funcitons?

Reply:
out of scope for this pr